modules/sdp/LocalSdpMunger.js

```diff
@@ -203,9 +203,29 @@ export default class LocalSdpMunger {
         const msid = mediaSection.ssrcs.find(s => s.attribute === 'msid');
 
         if (!this.tpc.isP2P
-            && (!msid || mediaSection.direction === 'recvonly' || mediaSection.direction === 'inactive')) {
+            && (!msid
+                || mediaSection.mLine?.direction === 'recvonly'
+                || mediaSection.mLine?.direction === 'inactive')) {
             mediaSection.ssrcs = undefined;
             mediaSection.ssrcGroups = undefined;
+
+        // Add the msid attribute if it is missing for p2p sources. Firefox doesn't produce a a=ssrc line
+        // with msid attribute.
+        } else if (this.tpc.isP2P && mediaSection.mLine?.direction === 'sendrecv') {
+            const msidLine = mediaSection.mLine?.msid;
+            const trackId = msidLine && msidLine.split(' ')[1];
+            const sources = [ ...new Set(mediaSection.mLine?.ssrcs?.map(s => s.id)) ];
+
+            for (const source of sources) {
+                const msidExists = mediaSection.ssrcs
+                    .find(ssrc => ssrc.id === source && ssrc.attribute === 'msid');
+
+                !msidExists && mediaSection.ssrcs.push({
+                    id: source,
+                    attribute: 'msid',
+                    value: `${this.localEndpointId}-${mediaSection.mLine?.type}-${pcId} ${trackId}-${pcId}`
```

Comment:
can we extract msid generation into a function? it seems to be duplicated with the logic at line 190?